File: android/app/src/main/java/com/mindful/android/models/RestrictionGroup.java
/*
 *
 *  *
 *  *  * Copyright (c) 2024 Mindful (https://github.com/akaMrNagar/Mindful)
 *  *  * Author : Pawan Nagar (https://github.com/akaMrNagar)
 *  *  *
 *  *  * This source code is licensed under the GPL-2.0 license license found in the
 *  *  * LICENSE file in the root directory of this source tree.
 *  *
 *
 */

package com.mindful.android.models;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashSet;

public class RestrictionGroup {

    /**
     * Unique ID for each restriction group
     */
    public final int id;

    /**
     * Name of the group
     */
    public final String groupName;

    /**
     * The timer set for the group in SECONDS
     */
    public final int timerSec;

    /**
     * [TimeOfDay] in minutes from where the active period will start.
     * It is stored as total minutes.
     */
    public final int activePeriodStart;

    /**
     * [TimeOfDay] in minutes when the active period will end
     * It is stored as total minutes.
     */
    public final int activePeriodEnd;

    /**
     * Total duration of active period from start to end in MINUTES
     */
    public final int periodDurationInMins;

    /**
     * List of app's packages which are associated with the group.
     */
    public final HashSet<String> distractingApps;


    /**
     * Constructor to initialize from JSON.
     *
     * @param jsonObject the JSON object containing restriction group data.
     */
    public RestrictionGroup(@NonNull JSONObject jsonObject) throws JSONException {
        this.id = jsonObject.optInt("id", 0);
        this.groupName = jsonObject.optString("groupName", "");
        this.timerSec = jsonObject.optInt("timerSec", 0);
        this.activePeriodStart = jsonObject.optInt("activePeriodStart", 0);
        this.activePeriodEnd = jsonObject.optInt("activePeriodEnd", 0);
        this.periodDurationInMins = jsonObject.optInt("periodDurationInMins", 0);

        // Deserialize apps package
        JSONArray appsJsonArray = jsonObject.optJSONArray("distractingApps");
        HashSet<String> distractingApps = new HashSet<>(0);
        if (appsJsonArray != null) {
            for (int i = 0; i < appsJsonArray.length(); i++) {
                distractingApps.add(appsJsonArray.getString(i));
            }
        }

        this.distractingApps = distractingApps;
    }
}

File: android/app/src/main/java/com/mindful/android/models/RestrictionGroup.kt
package com.mindful.android.models

import com.mindful.android.utils.JsonUtils
import org.json.JSONObject

/**
 * Represents a group of restricted apps with shared usage settings like timers and active periods.
 */
data class RestrictionGroup(
    /**
     * Unique ID of the group.
     */
    val id: Int = 0,

    /**
     * Name of the group (e.g., "Social", "Games").
     */
    val groupName: String = "Social",

    /**
     * Time limit for all apps in the group, in seconds.
     */
    val timerSec: Int = 0,

    /**
     * Time of day (in minutes) from midnight when usage is allowed to start.
     */
    val activePeriodStart: Int = 0,

    /**
     * Total duration (in minutes) of the allowed usage period.
     */
    val periodDurationInMins: Int = 0,

    /**
     * Set of app package names associated with this group.
     */
    val distractingApps: Set<String> = emptySet(),
) {
    companion object {

        /**
         * Constructs a [RestrictionGroup] from a JSON string.
         */
        fun fromJson(json: String): RestrictionGroup {
            val jsonObject = JSONObject(json)
            return RestrictionGroup(
                id = jsonObject.optInt("id", 0),
                groupName = jsonObject.optString("groupName", "Social"),
                timerSec = jsonObject.optInt("timerSec", 0),
                activePeriodStart = jsonObject.optInt("activePeriodStart", 0),
                periodDurationInMins = jsonObject.optInt("periodDurationInMins", 0),
                distractingApps = JsonUtils.parseStringSet(
                    jsonObject.optJSONArray("distractingApps")?.toString()
                )
            )
        }
    }
}
